torch/csrc/autograd/custom_function.cpp

```diff
@@ -26,6 +29,149 @@ Variable VariableInfo::zeros(at::OptionalDeviceGuard& device_guard) const {
   }
 }
 
+void _foreach_tensor(
+    const std::function<void (size_t, size_t, const at::Tensor&)>& fn,
+    torch::jit::Stack* stack,
+    size_t stack_start,
+    size_t size) {
+  // Enumerate over tensors in a stack, including ones in TensorLists
+  int idx = 0;
+  for (const auto i : c10::irange(size)) {
+    auto& ivalue = (*stack)[stack_start + i];
+    if (ivalue.isTensor()) {  // true for optional tensor that has value
+      const auto& tensor = ivalue.toTensor();
+      fn(idx, i, tensor);
+      idx++;
+    } if (ivalue.isTensorList()) {
+      for (const auto& tensor : ivalue.toTensorVector()) {
+        fn(idx, i, tensor);
+        idx++;
+      }
+    }
+  }
+}
+
+void autogradNotImplementedFallback(const c10::OperatorHandle& op, DispatchKeySet dispatch_keys, torch::jit::Stack* stack) {
+  // Mimics the logic of a VariableType NotImplemented kernel
+  const auto& schema = op.schema();
+  const auto& op_name = schema.operator_name().name;
+  const auto& arguments = schema.arguments();
+  const auto& returns = schema.returns();
+  const auto num_arguments = arguments.size();
+  const auto num_returns = returns.size();
+  const auto stack_start = stack->size() - num_arguments;
+  const bool grad_mode = GradMode::is_enabled();
+  std::vector<at::Tensor> tensors_requiring_grad;
```

Comment:
This doesn't need to contain owning Tensors; TensorRef is OK as you set edges prior to redispatching (which then destroys the ivalues on the stack)

Reply:
TensorRef doesn't quite work with vectors. Might have to use a vector of Tensor ptr here...